tests/tests/class-plugin-functions.php

```diff
@@ -0,0 +1,106 @@
+	}
+
+	/**
+	 * Test blacklisted commands
+	 */
+	function test_blacklist_using_is_command_allowed() {
+		$this->assertFalse( WP_CLI_Cron_Control_Offload\is_command_allowed( 'cli' ) );
+	}
+
+	/**
+	 * Test whitelisted command validation
+	 */
+	function test_whitelist_using_validate_command() {
+		$this->assertTrue( is_string( WP_CLI_Cron_Control_Offload\validate_command( 'wp post list' ) ) );
+		$this->assertTrue( is_string( WP_CLI_Cron_Control_Offload\validate_command( 'post list' ) ) );
+	}
+
+	/**
+	 * Test blacklisted command validation
+	 */
+	function test_blacklist_using_validate_command() {
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'wp cli info' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'cli info' ) ) );
+	}
+
+	/**
+	 * Test scheduling several of the same allowed event
+	 */
+	function test_allowed_event_scheduling() {
+		// Should succeed, returning a timestamp.
+		$this->assertTrue( is_int( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'wp post list' ) ) );
+
+		// Should be blocked as a duplicate, thanks to Core's 10-minute lookahead.
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'wp post list' ) ) );
+
+		// Should also fail as normalization makes it a duplicate.
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'post list' ) ) );
+	}
+
+	/**
+	 * Test scheduling several of the same blocked event
+	 */
+	function test_blocked_event_scheduling() {
+		// Should fail, is a blocked event.
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'wp cli info' ) ) );
+
+		// Should fail as a blocked event, would otherwise fail as a duplicate.
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'wp cli info' ) ) );
+
+		// Should also fail as a blocked event, though normalization would also block it as a duplicate.
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\schedule_cli_command( 'cli info' ) ) );
+	}
+
+	/**
+	 * Test each blocked bash operator
+	 */
+	function test_for_invalid_bash_operators() {
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list & date' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list | date' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list > /tmp/nope' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list 2> /tmp/nope' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list 1>&2 /tmp/nope' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list 2>&1 /tmp/nope' ) ) );
+		$this->assertTrue( is_wp_error( WP_CLI_Cron_Control_Offload\validate_command( 'post list &> /tmp/nope' ) ) );
```

Comment:
For tests like this when we are running through a list of inputs/outputs, it's better to use PHPUnit's data providers: https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers -- each input gets executed as a separate test and we can label them so we know which ones fail.

Same thing applies to `test_blocked_event_scheduling` and a bunch of other ones above where the main action and assertion between each test is the same.

Finally, it's nice to use this pattern for tests: http://wiki.c2.com/?ArrangeActAssert -- i.e. instead of squeezing everything into one line, opt for readability instead:

```
$input = 'post list & date';

$output = WP_CLI_Cron_Control_Offload\validate_command( $input );

$this->assertWpError( $output );
```